File: resmon_scripts/frontend/src/api/evidence.ts
import { getBaseUrl } from './client';
import { object, uuid, LibraryFile } from './library';

export type EvidenceFile = Pick<LibraryFile, 'file_id'|'version_id'|'vault_id'|'sha256'|'byte_size'|'media_type'|'original_name'|'created_at_utc'>;
export interface Project {id:number; project_id:string; vault_id:string; name:string; revision:number; created_at_utc:string; updated_at_utc:string; file_count:number; note_count:number}
export interface Member {id:number; project_id:string; file_id:string; version_id:string; added_at_utc:string; availability:'not_checked'; file:EvidenceFile}
export interface Anchor {page_number:number; extraction_contract:string; page_text_sha256:string; start_codepoint:number; end_codepoint:number; quote:string}
export interface SavedNote {id:number; note_id:string; project_id:string; file_id:string; version_id:string; kind:'note'|'passage'; body:string; revision:number; created_at_utc:string; updated_at_utc:string; file:EvidenceFile; membership_state:'member'|'removed'; resolution:'not_checked'; page_number:number|null; extraction_contract:string|null; page_text_sha256:string|null; start_codepoint:number|null; end_codepoint:number|null; quote:string|null}
export interface Page<T> {items:T[]; through_id:number; next_after_id:number|null; has_more:boolean; total:number; count_basis:string; revision?:number}
export interface TextPage {contract_version:1; vault_id:string; project_id:string; file_id:string; version_id:string; sha256:string; media_type:string; status:'extracted'|'no_text'|'unsupported'|'malformed'|'limit_exceeded'|'timeout'|'unavailable'; page_number:number; page_count:number|null; text:string; extraction_contract:string; page_text_sha256:string|null; examined_pages:number[]; remaining_pages:'not_examined'; coverage:string}
export const integer=(x:unknown,min=0,max=Number.MAX_SAFE_INTEGER):x is number=>Number.isSafeInteger(x)&&(x as number)>=min&&(x as number)<=max;
const literal=(x:unknown,min:number,max:number):x is string=>typeof x==='string'&&Array.from(x).length>=min&&Array.from(x).length<=max&&!/[\x00\uD800-\uDFFF]/u.test(x);
const digest=(x:unknown):x is string=>typeof x==='string'&&/^[a-f0-9]{64}$/.test(x);
const stamp=(x:unknown):x is string=>typeof x==='string'&&/^\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d(?:\.\d{1,6})?(?:\+00:00|Z)$/.test(x);
const fail=():never=>{throw new Error('Evidence response does not match the selected identity or bounds. Refresh explicitly.');};
export function validateFile(raw:unknown,vault:string):EvidenceFile {
  const x=object(raw);
  if(!uuid(vault)||x.vault_id!==vault||!uuid(x.file_id)||!uuid(x.version_id)||!digest(x.sha256)||!integer(x.byte_size,1,67108864)||!['application/pdf','text/plain','text/markdown'].includes(String(x.media_type))||!literal(x.original_name,1,255)||!stamp(x.created_at_utc))fail();
  return x as unknown as EvidenceFile;
}
export function validateProject(raw:unknown,vault:string,id?:string):Project {
  const x=object(raw);
  if(!integer(x.id,1)||!uuid(x.project_id)||x.vault_id!==vault||(id&&x.project_id!==id)||!literal(x.name,1,120)||!integer(x.revision,1)||!integer(x.file_count,0,1000)||!integer(x.note_count,0,5000)||!stamp(x.created_at_utc)||!stamp(x.updated_at_utc))fail();
  return x as unknown as Project;
}
export function validateNote(raw:unknown,vault:string,project:string):SavedNote {
  const x=object(raw);const f=validateFile(x.file,vault);
  if(!integer(x.id,1)||!uuid(x.note_id)||x.project_id!==project||x.file_id!==f.file_id||x.version_id!==f.version_id||!integer(x.revision,1)||!stamp(x.created_at_utc)||!stamp(x.updated_at_utc)||!['member','removed'].includes(String(x.membership_state))||x.resolution!=='not_checked'||!literal(x.body,x.kind==='note'?1:0,20000))fail();
  const fields=['page_number','extraction_contract','page_text_sha256','start_codepoint','end_codepoint','quote'];
  if(x.kind==='note'){if(fields.some(k=>x[k]!==null))fail();}
  else if(x.kind==='passage'){
    if(!integer(x.page_number,1,200)||!literal(x.extraction_contract,1,100)||!digest(x.page_text_sha256)||!integer(x.start_codepoint)||!integer(x.end_codepoint,1)||!literal(x.quote,1,20000)||(x.end_codepoint as number)-(x.start_codepoint as number)!==Array.from(x.quote as string).length)fail();
  }else fail();
  return x as unknown as SavedNote;
}
function envelope(raw:unknown,vault:string,project?:string):Record<string,unknown>{const x=object(raw);if(!uuid(vault)||x.contract_version!==1||x.vault_id!==vault||(project&&(!uuid(project)||x.project_id!==project)))fail();return x;}
const query=(values:Record<string,string|number|undefined>)=>'?'+new URLSearchParams(Object.entries(values).filter(([,v])=>v!==undefined).map(([k,v])=>[k,String(v)])).toString();
export function evidenceBaseUrl():string {
  const port=window.resmonAPI?.getBackendPort();
  if(!port||!/^([1-9][0-9]{0,4})$/.test(port)||Number(port)>65535||port==='8742')throw new Error('An explicitly identified desktop backend is required for Evidence. No default instance is selected.');
  const base=getBaseUrl();if(base!==`http://127.0.0.1:${port}`)throw new Error('The selected backend changed.');return base;
}
export async function evidenceRequest(path:string,method='GET',body?:unknown,signal?:AbortSignal):Promise<Response>{
  const response=await fetch(`${evidenceBaseUrl()}/api/evidence${path}`,{method,signal,cache:'no-store',headers:{'X-Resmon-Library':'1',...(body!==undefined?{'Content-Type':'application/json'}:{})},...(body!==undefined?{body:JSON.stringify(body)}:{})});
  if(!response.ok){let message=`Evidence request refused (${response.status}).`;try{const x=object(await response.json());const d=x.detail;if(typeof d==='string')message=d;else if(d&&typeof d==='object'&&typeof object(d).message==='string')message=String(object(d).message);}catch{/* Keep redacted status. */}throw new Error(message);}
  return response;
}
const read=async(path:string,method='GET',body?:unknown,signal?:AbortSignal):Promise<unknown>=>(await evidenceRequest(path,method,body,signal)).json();
const base=(p:Project)=>({expected_vault_id:p.vault_id,expected_revision:p.revision});
function projectResult(raw:unknown,vault:string,id?:string):Project{const x=envelope(raw,vault,id);const p=validateProject(x.project,vault,id);if(x.project_id!==p.project_id)fail();return p;}
function pageResult<T extends {id:number}>(x:Record<string,unknown>,key:string,validate:(raw:unknown)=>T,through?:number,after=0):Page<T>{
  if(!Array.isArray(x[key])||(x[key] as unknown[]).length>50||!integer(x.through_id)||(through!==undefined&&x.through_id!==through)||!integer(x.total)||typeof x.count_basis!=='string'||typeof x.has_more!=='boolean'||!(x.next_after_id===null||integer(x.next_after_id,1))||(x.revision!==undefined&&!integer(x.revision,1)))fail();
  const items=(x[key] as unknown[]).map(validate);
  if(items.some((it,i)=>it.id<=after||it.id>(x.through_id as number)||(i>0&&it.id<=items[i-1].id))||(x.has_more?items.length===0||x.next_after_id!==items[items.length-1].id:x.next_after_id!==null))fail();
  return {items,through_id:x.through_id as number,next_after_id:x.next_after_id as number|null,has_more:x.has_more as boolean,total:x.total as number,count_basis:x.count_basis as string,revision:x.revision as number|undefined};
}
export async function sha256(bytes:Uint8Array):Promise<string>{const copy=new Uint8Array(bytes);return Array.from(new Uint8Array(await crypto.subtle.digest('SHA-256',copy.buffer)),b=>b.toString(16).padStart(2,'0')).join('');}
export async function validateText(raw:unknown,p:Project,f:EvidenceFile,page:number):Promise<TextPage>{
  const x=envelope(raw,p.vault_id,p.project_id);const expected=f.media_type==='application/pdf'?'pypdf-6.18.1/plain-lf/v1':'library-text-lf/v1';
  if(x.file_id!==f.file_id||x.version_id!==f.version_id||x.sha256!==f.sha256||x.media_type!==f.media_type||x.page_number!==page||!['extracted','no_text','unsupported','malformed','limit_exceeded','timeout','unavailable'].includes(String(x.status))||!literal(x.text,0,f.media_type==='application/pdf'?200000:262144)||/[\r\x00]/.test(x.text as string)||x.extraction_contract!==expected||!(x.page_count===null||integer(x.page_count,0))||!Array.isArray(x.examined_pages)||x.remaining_pages!=='not_examined'||typeof x.coverage!=='string'||new TextEncoder().encode(JSON.stringify(x)).length>2097152)fail();
  if(f.media_type!=='application/pdf'&&((x.text as string).split('\n').length>5000||f.byte_size>262144||page!==1))fail();
  if(x.status==='extracted'||x.status==='no_text'){
    if(!integer(x.page_count,page,200)||JSON.stringify(x.examined_pages)!==JSON.stringify([page])||(x.status==='extracted')!==Boolean(x.text))fail();
    if(x.status==='extracted'){if(!digest(x.page_text_sha256)||await sha256(new TextEncoder().encode(x.text as string))!==x.page_text_sha256)fail();}else if(x.page_text_sha256!==null)fail();
  }else if(x.text!==''||x.page_text_sha256!==null||(x.examined_pages as unknown[]).length!==0)fail();
  return x as unknown as TextPage;
}
export const evidenceApi={
  projects:async(vault:string,through?:number,after?:number,signal?:AbortSignal)=>pageResult(envelope(await read('/projects'+query({expected_vault_id:vault,through_id:through,after_id:after}),undefined,undefined,signal),vault),'projects',x=>validateProject(x,vault),through,after),
  create:async(vault:string,name:string)=>projectResult(await read('/projects','POST',{expected_vault_id:vault,name}),vault),
  detail:async(vault:string,id:string,signal?:AbortSignal)=>projectResult(await read(`/projects/${id}`+query({expected_vault_id:vault}),undefined,undefined,signal),vault,id),
  rename:async(p:Project,name:string)=>projectResult(await read(`/projects/${p.project_id}`,'PATCH',{...base(p),name}),p.vault_id,p.project_id),
  files:async(p:Project,through?:number,after?:number,signal?:AbortSignal)=>pageResult(envelope(await read(`/projects/${p.project_id}/files`+query({expected_vault_id:p.vault_id,through_id:through,after_id:after}),undefined,undefined,signal),p.vault_id,p.project_id),'files',raw=>{const x=object(raw);const f=validateFile(x.file,p.vault_id);if(!integer(x.id,1)||x.project_id!==p.project_id||x.file_id!==f.file_id||x.version_id!==f.version_id||x.availability!=='not_checked'||!stamp(x.added_at_utc))fail();return x as unknown as Member;},through,after),
  add:async(p:Project,f:EvidenceFile)=>{const x=envelope(await read(`/projects/${p.project_id}/files`,'POST',{...base(p),file_id:f.file_id,version_id:f.version_id}),p.vault_id,p.project_id);const m=object(x.membership);if(m.project_id!==p.project_id||m.file_id!==f.file_id||m.version_id!==f.version_id||typeof x.added!=='boolean')fail();return projectResult(x,p.vault_id,p.project_id);},
  remove:async(p:Project,f:EvidenceFile)=>{const x=envelope(await read(`/projects/${p.project_id}/files/${f.file_id}`,'DELETE',{...base(p),version_id:f.version_id}),p.vault_id,p.project_id);if(typeof x.removed!=='boolean')fail();return projectResult(x,p.vault_id,p.project_id);},
  notes:async(p:Project,through?:number,after?:number,signal?:AbortSignal)=>pageResult(envelope(await read(`/projects/${p.project_id}/notes`+query({expected_vault_id:p.vault_id,through_id:through,after_id:after}),undefined,undefined,signal),p.vault_id,p.project_id),'notes',x=>validateNote(x,p.vault_id,p.project_id),through,after),
  save:async(p:Project,f:EvidenceFile,body:string,anchor?:Anchor)=>{const x=envelope(await read(`/projects/${p.project_id}/notes`,'POST',{...base(p),file_id:f.file_id,version_id:f.version_id,kind:anchor?'passage':'note',body,anchor:anchor??null}),p.vault_id,p.project_id);const note=validateNote(x.note,p.vault_id,p.project_id);if(note.file_id!==f.file_id||note.version_id!==f.version_id||note.body!==body||note.kind!==(anchor?'passage':'note'))fail();if(anchor&&Object.entries(anchor).some(([k,v])=>object(note)[k]!==v))fail();return {project:projectResult(x,p.vault_id,p.project_id),note};},
  edit:async(p:Project,n:SavedNote,body:string)=>{const x=envelope(await read(`/projects/${p.project_id}/notes/${n.note_id}`,'PATCH',{...base(p),expected_note_revision:n.revision,body}),p.vault_id,p.project_id);const note=validateNote(x.note,p.vault_id,p.project_id);if(note.note_id!==n.note_id||note.body!==body||note.revision!==n.revision+1||['file_id','version_id','kind','page_number','extraction_contract','page_text_sha256','start_codepoint','end_codepoint','quote'].some(k=>object(note)[k]!==object(n)[k]))fail();return {project:projectResult(x,p.vault_id,p.project_id),note};},
  text:async(p:Project,f:EvidenceFile,page:number,signal?:AbortSignal)=>validateText(await read(`/projects/${p.project_id}/reader/${f.file_id}`+query({expected_vault_id:p.vault_id,version_id:f.version_id,page,representation:'text'}),undefined,undefined,signal),p,f,page),
  pdf:async(p:Project,f:EvidenceFile,signal?:AbortSignal)=>{const r=await evidenceRequest(`/projects/${p.project_id}/reader/${f.file_id}`+query({expected_vault_id:p.vault_id,version_id:f.version_id,representation:'pdf'}),'GET',undefined,signal);const headers={'X-Resmon-Evidence-Contract':'1','X-Resmon-Vault':p.vault_id,'X-Resmon-Project':p.project_id,'X-Resmon-File':f.file_id,'X-Resmon-Version':f.version_id,'X-Resmon-SHA256':f.sha256};if(Object.entries(headers).some(([k,v])=>r.headers.get(k)!==v)||f.byte_size>16777216)fail();const data=new Uint8Array(await boundedBytes(r,16777216));if(data.length!==f.byte_size||await sha256(data)!==f.sha256)fail();return data;},
};
export async function boundedBytes(response:Response,maximum:number):Promise<ArrayBuffer>{
  const reader=response.body?.getReader();if(!reader)throw new Error('A bounded response stream is unavailable.');const chunks:Uint8Array[]=[];let size=0;
  try{while(true){const {value,done}=await reader.read();if(done)break;size+=value.byteLength;if(size>maximum)throw new Error('Download exceeds its stated limit.');chunks.push(value);}}catch(error){await reader.cancel();throw error;}finally{reader.releaseLock();}
  const bytes=new Uint8Array(size);let offset=0;for(const chunk of chunks){bytes.set(chunk,offset);offset+=chunk.length;}return bytes.buffer;
}

File: resmon_scripts/implementation_scripts/evidence.py
"""Exact Library versions, collection membership and literal saved evidence.

These records are owner selections, not inferred paper identity or scientific
findings. Removing membership never removes a saved record or retained object.
"""
from __future__ import annotations

from contextlib import contextmanager
from datetime import datetime, timezone
import re
import sqlite3
import uuid
from typing import Iterator

from . import library

MAX_PROJECTS = 100
MAX_MEMBERS = 1000
MAX_NOTES = 5000
MAX_BODY = 20000
FILE_FIELDS = ('file_id', 'version_id', 'vault_id', 'sha256', 'byte_size',
               'media_type', 'original_name', 'created_at_utc')
ANCHOR_FIELDS = ('page_number', 'extraction_contract', 'page_text_sha256',
                 'start_codepoint', 'end_codepoint', 'quote')
TEXT_CONTRACT = 'library-text-lf/v1'
PDF_CONTRACT = 'pypdf-6.18.1/plain-lf/v1'


def utc_now() -> str:
    return datetime.now(timezone.utc).isoformat(timespec='microseconds')


class EvidenceError(library.LibraryError):
    pass


def refuse(reason: str, message: str, status: int = 409) -> None:
    raise EvidenceError(reason, message, status)


def identity(value: str) -> str:
    try:
        return library.identity(value)
    except library.LibraryError:
        refuse('invalid_identity', 'An exact canonical UUID is required.', 422)


def integer(value: int, minimum: int = 1, maximum: int = 2**53 - 1) -> int:
    if type(value) is not int or not minimum <= value <= maximum:
        refuse('invalid_integer', 'An integer within the stated bounds is required.', 422)
    return value


def literal(value: str, minimum: int, maximum: int) -> str:
    if (not isinstance(value, str) or not minimum <= len(value) <= maximum
            or '\x00' in value or any(0xD800 <= ord(c) <= 0xDFFF for c in value)):
        refuse('invalid_text', 'Literal text exceeds its bounds or contains invalid characters.', 422)
    return value


def timestamp(value: str) -> str:
    # Evidence timestamps are generated here at six-digit UTC precision. A
    # damaged stored record must not become apparently valid portable evidence.
    try:
        if (not isinstance(value, str)
                or not re.fullmatch(r'\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d\.\d{6}\+00:00', value)):
            raise ValueError
        datetime.fromisoformat(value)
    except ValueError:
        refuse('corrupt_record', 'A stored Evidence timestamp is invalid.')
    return value


def name(value: str) -> str:
    if not isinstance(value, str):
        refuse('invalid_name', 'A project name is required.', 422)
    return literal(value.strip(), 1, 120)


def vault(conn: sqlite3.Connection, expected: str) -> str:
    identity(expected)
    row = conn.execute('SELECT vault_id FROM library_vault WHERE singleton=1').fetchone()
    if row is None:
        refuse('no_vault', 'Configure Library before creating a collection.')
    if row[0] != expected:
        refuse('wrong_vault', 'The configured Library identity changed. Refresh explicitly.')
    return expected


def envelope(vault_id: str, project_id: str | None = None, **fields: object) -> dict:
    return {'contract_version': 1, 'vault_id': vault_id, **({'project_id': project_id} if project_id else {}), **fields}


def project(conn: sqlite3.Connection, expected: str, project_id: str) -> dict:
    vault(conn, expected)
    identity(project_id)
    row = conn.execute('SELECT * FROM evidence_projects WHERE project_id=?', (project_id,)).fetchone()
    if row is None:
        refuse('unknown_project', 'This project is not present.', 404)
    row = dict(row)
    if row['vault_id'] != expected:
        refuse('wrong_vault', 'This project belongs to a different vault.')
    integer(row['id']); integer(row['revision']); literal(row['name'], 1, 120)
    timestamp(row['created_at_utc']); timestamp(row['updated_at_utc'])
    return row


def project_view(conn: sqlite3.Connection, row: dict) -> dict:
    counts = {key: conn.execute(f'SELECT count(*) FROM {table} WHERE project_id=?', (row['project_id'],)).fetchone()[0]
              for key, table in [('file_count', 'evidence_project_files'), ('note_count', 'evidence_notes')]}
    return {**row, **counts}


def detail(conn: sqlite3.Connection, expected: str, project_id: str) -> dict:
    row = project(conn, expected, project_id)
    return envelope(expected, project_id, project=project_view(conn, row))


def file_record(conn: sqlite3.Connection, expected: str, file_id: str, version_id: str) -> dict:
    identity(file_id); identity(version_id)
    row = library.file_row(conn, file_id, version_id)
    if row['vault_id'] != expected:
        refuse('wrong_vault', 'The file and project vault do not match.')
    if not isinstance(row['sha256'], str) or not re.fullmatch('[0-9a-f]{64}', row['sha256']):
        refuse('corrupt_record', 'The recorded digest is invalid.')
    integer(row['byte_size'], 1, library.MAX_FILE_BYTES)
    if row['media_type'] not in library.MEDIA.values():
        refuse('corrupt_record', 'The recorded file type is invalid.')
    literal(row['original_name'], 1, 255)
    return {k: row[k] for k in FILE_FIELDS}


def member(conn: sqlite3.Connection, expected: str, project_id: str, file_id: str, version_id: str,
           *, required: bool = True) -> dict | None:
    file_record(conn, expected, file_id, version_id)
    row = conn.execute('SELECT * FROM evidence_project_files WHERE project_id=? AND file_id=?', (project_id, file_id)).fetchone()
    if row is None:
        if required:
            refuse('not_member', 'Select a current member of this collection.')
        return None
    if row['version_id'] != version_id:
        refuse('wrong_version', 'Collection membership and file version do not match.')
    integer(row['id']); timestamp(row['added_at_utc'])
    return dict(row)


@contextmanager
def write(conn: sqlite3.Connection, expected: str, project_id: str,
          expected_revision: int) -> Iterator[dict]:
    integer(expected_revision)
    conn.execute('BEGIN IMMEDIATE')
    try:
        row = project(conn, expected, project_id)
        if row['revision'] != expected_revision:
            refuse('stale_revision', 'This project changed. Refresh; your unsaved text is preserved.')
        yield row
        conn.commit()
    except BaseException:
        conn.rollback()
        raise


def changed(conn: sqlite3.Connection, project_id: str) -> None:
    conn.execute('UPDATE evidence_projects SET revision=revision+1,updated_at_utc=? WHERE project_id=?',
                 (utc_now(), project_id))


def create_project(conn: sqlite3.Connection, expected: str, title: str) -> dict:
    title = name(title)
    conn.execute('BEGIN IMMEDIATE')
    try:
        vault(conn, expected)
        if conn.execute('SELECT count(*) FROM evidence_projects').fetchone()[0] >= MAX_PROJECTS:
            refuse('project_limit', 'The 100-project limit is reached.', 413)
        project_id, now = str(uuid.uuid4()), utc_now()
        conn.execute('INSERT INTO evidence_projects(project_id,vault_id,name,revision,created_at_utc,updated_at_utc) VALUES (?,?,?,1,?,?)',
                     (project_id, expected, title, now, now))
        result = detail(conn, expected, project_id)
        conn.commit()
        return result
    except BaseException:
        conn.rollback()
        raise


def rename_project(conn: sqlite3.Connection, expected: str, project_id: str, revision: int, title: str) -> dict:
    title = name(title)
    with write(conn, expected, project_id, revision):
        conn.execute('UPDATE evidence_projects SET name=? WHERE project_id=?', (title, project_id))
        changed(conn, project_id)
        return detail(conn, expected, project_id)


def pagination(conn: sqlite3.Connection, table: str, where: str, args: tuple,
               after_id: int = 0, through_id: int | None = None, limit: int = 50) -> tuple[list, dict]:
    integer(after_id, 0); integer(limit, 1, 50)
    if through_id is not None:
        integer(through_id, 0)
    ceiling = through_id if through_id is not None else conn.execute(f'SELECT coalesce(max(id),0) FROM {table} WHERE {where}', args).fetchone()[0]
    rows = conn.execute(f'SELECT * FROM {table} WHERE {where} AND id>? AND id<=? ORDER BY id LIMIT ?', (*args, after_id, ceiling, limit + 1)).fetchall()
    more = len(rows) > limit
    items = [dict(row) for row in rows[:limit]]
    total = conn.execute(f'SELECT count(*) FROM {table} WHERE {where} AND id<=?', (*args, ceiling)).fetchone()[0]
    return items, {'through_id': ceiling, 'next_after_id': items[-1]['id'] if more else None,
                   'has_more': more, 'total': total, 'count_basis': 'current rows at the selected ID ceiling; removals may change counts'}


def list_projects(conn: sqlite3.Connection, expected: str, **page: int) -> dict:
    vault(conn, expected)
    rows, bounds = pagination(conn, 'evidence_projects', 'vault_id=?', (expected,), **page)
    items = [project_view(conn, project(conn, expected, r['project_id'])) for r in rows]
    return envelope(expected, projects=items, **bounds)


def list_files(conn: sqlite3.Connection, expected: str, project_id: str, **page: int) -> dict:
    p = project(conn, expected, project_id)
    rows, bounds = pagination(conn, 'evidence_project_files', 'project_id=?', (project_id,), **page)
    for r in rows:
        integer(r['id']); timestamp(r['added_at_utc'])
    items = [{**r, 'file': file_record(conn, expected, r['file_id'], r['version_id']), 'availability': 'not_checked'} for r in rows]
    return envelope(expected, project_id, revision=p['revision'], files=items, **bounds)


def add_file(conn: sqlite3.Connection, expected: str, project_id: str, revision: int, file_id: str, version_id: str) -> dict:
    with write(conn, expected, project_id, revision):
        current = member(conn, expected, project_id, file_id, version_id, required=False)
        if current is None:
            if conn.execute('SELECT count(*) FROM evidence_project_files WHERE project_id=?', (project_id,)).fetchone()[0] >= MAX_MEMBERS:
                refuse('member_limit', 'This project has reached 1,000 files.', 413)
            conn.execute('INSERT INTO evidence_project_files(project_id,file_id,version_id,added_at_utc) VALUES (?,?,?,?)',
                         (project_id, file_id, version_id, utc_now()))
            changed(conn, project_id)
        return envelope(expected, project_id, project=project_view(conn, project(conn, expected, project_id)),
                        membership=member(conn, expected, project_id, file_id, version_id), added=current is None)


def remove_file(conn: sqlite3.Connection, expected: str, project_id: str, revision: int, file_id: str, version_id: str) -> dict:
    with write(conn, expected, project_id, revision):
        current = member(conn, expected, project_id, file_id, version_id, required=False)
        if current is not None:
            conn.execute('DELETE FROM evidence_project_files WHERE project_id=? AND file_id=?', (project_id, file_id))
            changed(conn, project_id)
        return envelope(expected, project_id, project=project_view(conn, project(conn, expected, project_id)), removed=current is not None)


def validate_note(row: dict) -> None:
    if 'id' in row:
        integer(row['id'])
    timestamp(row['created_at_utc']); timestamp(row['updated_at_utc'])
    for key in ('note_id', 'project_id', 'file_id', 'version_id'):
        identity(row[key])
    integer(row['revision']); literal(row['body'], 0 if row['kind'] == 'passage' else 1, MAX_BODY)
    if row['kind'] == 'note':
        if any(row[k] is not None for k in ANCHOR_FIELDS):
            refuse('corrupt_record', 'A plain note has an invalid anchor.')
    elif row['kind'] == 'passage':
        integer(row['page_number'], 1, 200); integer(row['start_codepoint'], 0); integer(row['end_codepoint'])
        literal(row['extraction_contract'], 1, 100)
        literal(row['quote'], 1, MAX_BODY)
        if (row['end_codepoint'] - row['start_codepoint'] != len(row['quote'])
                or not isinstance(row['extraction_contract'], str)
                or not isinstance(row['page_text_sha256'], str)
                or not re.fullmatch('[0-9a-f]{64}', row['page_text_sha256'])):
            refuse('corrupt_record', 'The saved passage basis is invalid.')
    else:
        refuse('corrupt_record', 'The saved note kind is invalid.')


def note_view(conn: sqlite3.Connection, expected: str, row: dict) -> dict:
    validate_note(row)
    f = file_record(conn, expected, row['file_id'], row['version_id'])
    m = member(conn, expected, row['project_id'], row['file_id'], row['version_id'], required=False)
    return {**row, 'file': f, 'membership_state': 'member' if m else 'removed', 'resolution': 'not_checked'}


def list_notes(conn: sqlite3.Connection, expected: str, project_id: str, file_id: str | None = None,
               version_id: str | None = None, **page: int) -> dict:
    p = project(conn, expected, project_id)
    where, args = 'project_id=?', (project_id,)
    if file_id is not None or version_id is not None:
        file_record(conn, expected, file_id, version_id)
        where += ' AND file_id=? AND version_id=?'; args += (file_id, version_id)
    rows, bounds = pagination(conn, 'evidence_notes', where, args, **page)
    return envelope(expected, project_id, revision=p['revision'], notes=[note_view(conn, expected, r) for r in rows], **bounds)


def create_note(conn: sqlite3.Connection, expected: str, project_id: str, revision: int,
                file_id: str, version_id: str, kind: str, body: str, anchor: dict | None = None,
                *, cancel: object = None) -> dict:
    project(conn, expected, project_id)
    member(conn, expected, project_id, file_id, version_id)
    literal(body, 1 if kind == 'note' else 0, MAX_BODY)
    values = {k: None for k in ANCHOR_FIELDS}
    if kind == 'passage':
        if not isinstance(anchor, dict) or set(anchor) != set(ANCHOR_FIELDS):
            refuse('invalid_anchor', 'A complete exact passage basis is required.', 422)
        values.update(anchor)
    elif kind != 'note' or anchor is not None:
        refuse('invalid_note', 'Choose a plain note or an exact passage.', 422)
    now = utc_now()
    row = {'note_id': str(uuid.uuid4()), 'project_id': project_id, 'file_id': file_id,
           'version_id': version_id, 'kind': kind, 'body': body, **values,
           'revision': 1, 'created_at_utc': now, 'updated_at_utc': now}
    validate_note(row)
    if kind == 'passage':
        from . import evidence_reader
        basis = evidence_reader.read_text(conn, expected, project_id, file_id, version_id, row['page_number'], cancel=cancel)
        if (basis['status'] != 'extracted' or basis['extraction_contract'] != row['extraction_contract']
                or basis['page_text_sha256'] != row['page_text_sha256']
                or row['end_codepoint'] > len(basis['text'])
                or basis['text'][row['start_codepoint']:row['end_codepoint']] != row['quote']):
            refuse('quote_mismatch', 'The exact page, text hash, offsets and quote must match the retained version.')
    if cancel is not None and cancel.is_set():
        refuse('cancelled', 'Saving the selected evidence was cancelled.')
    with write(conn, expected, project_id, revision):
        if cancel is not None and cancel.is_set():
            refuse('cancelled', 'Saving the selected evidence was cancelled.')
        member(conn, expected, project_id, file_id, version_id)
        if conn.execute('SELECT count(*) FROM evidence_notes WHERE project_id=?', (project_id,)).fetchone()[0] >= MAX_NOTES:
            refuse('note_limit', 'This project has reached 5,000 saved notes.', 413)
        keys = tuple(row)
        conn.execute(f'INSERT INTO evidence_notes({",".join(keys)}) VALUES ({",".join("?" for _ in keys)})', tuple(row[k] for k in keys))
        changed(conn, project_id)
        saved = dict(conn.execute('SELECT * FROM evidence_notes WHERE note_id=?', (row['note_id'],)).fetchone())
        return envelope(expected, project_id, project=project_view(conn, project(conn, expected, project_id)), note=note_view(conn, expected, saved))


def edit_note(conn: sqlite3.Connection, expected: str, project_id: str, revision: int,
              note_id: str, note_revision: int, body: str) -> dict:
    identity(note_id); integer(note_revision)
    with write(conn, expected, project_id, revision):
        row = conn.execute('SELECT * FROM evidence_notes WHERE project_id=? AND note_id=?', (project_id, note_id)).fetchone()
        if row is None:
            refuse('unknown_note', 'This saved note is not present.', 404)
        row = dict(row); note_view(conn, expected, row)
        if row['revision'] != note_revision:
            refuse('stale_note', 'This note changed. Refresh; your unsaved text is preserved.')
        literal(body, 1 if row['kind'] == 'note' else 0, MAX_BODY)
        conn.execute('UPDATE evidence_notes SET body=?,revision=revision+1,updated_at_utc=? WHERE note_id=?',
                     (body, utc_now(), note_id))
        changed(conn, project_id)
        saved = dict(conn.execute('SELECT * FROM evidence_notes WHERE note_id=?', (note_id,)).fetchone())
        return envelope(expected, project_id, project=project_view(conn, project(conn, expected, project_id)), note=note_view(conn, expected, saved))


def reader_access(conn: sqlite3.Connection, expected: str, project_id: str, file_id: str, version_id: str) -> dict:
    project(conn, expected, project_id)
    f = file_record(conn, expected, file_id, version_id)
    if member(conn, expected, project_id, file_id, version_id, required=False) is None:
        notes = conn.execute('SELECT * FROM evidence_notes WHERE project_id=? AND file_id=?', (project_id, file_id)).fetchall()
        if not notes:
            refuse('not_member', 'This file has neither membership nor a saved note in this project.')
        for note in notes:
            note_view(conn, expected, dict(note))
    return f
